File: LeetCode Problems/1 to 100/23. Merge k Sorted Lists.cpp
class Solution {
    ListNode *head;
public:
    void merge(int data) {
        ListNode *curr = head, *prev = nullptr, *newNode = new ListNode();
        while (curr->next != nullptr) {
            if (curr->val > data)
                break;
            prev = curr;
            curr = curr->next;
        }
        newNode->val = data;
        if (curr != nullptr && curr->val < data) {
            prev = curr;
            prev->next = newNode;
        } else {
            if (prev == nullptr) {
                newNode->next = curr;
                head = newNode;
            } else {
                newNode->next = curr;
                prev->next = newNode;
            }
        }        
    }
    ListNode* mergeKLists(vector<ListNode*>& lists) {
        for (int i = 0; i < lists.size(); i++) {
            if (head == nullptr) {
                head = lists[i];
            } else {
                ListNode *temp = lists[i];
                while(temp != nullptr) {
                    merge(temp->val);
                    temp = temp->next;
                }
            }
        }
        return head;
    }
};

File: LeetCode Problems/1 to 100/23. Merge k Sorted Lists/23. Merge k Sorted Lists.java
class Solution {
    public ListNode mergeKLists(ListNode[] lists) {
        if (lists.length == 0) return null;
        ListNode result = lists[0];
        int i = 1;
        while (i < lists.length) {
            result = mergeTwoLists(result, lists[i]);
            i++;
        }
        return result;
    }

    ListNode mergeTwoLists(ListNode list1, ListNode list2) {
        if (list1 == null) return list2;
        if (list2 == null) return list1;

        if (list1.val < list2.val) {
            list1.next = mergeTwoLists(list1.next, list2);
            return list1;
        } else {
            list2.next = mergeTwoLists(list1, list2.next);
            return list2;
        }
    }
}
